### PR TITLE
Add phpcs rule to detect unused variables; Fix existing issues

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
This PR adds a PHP rule to detect unused PHP variables. And fixes the problems detected by this rule.

The fact we have many existing problems is a small proof that this rule is useful.

## How has this been tested?
I executed `composer lint`.
I verified no lint error were detected.

### CODE REVIEW

File: phpcs.xml.dist [hunk 1]
```diff
@@ -10,6 +10,11 @@
 	<rule ref="WordPress.WP.I18n"/>
 	<config name="text_domain" value="gutenberg,default"/>
 
+	<rule ref="VariableAnalysis.CodeAnalysis.VariableAnalysis">
+		<properties>
+			<property name="allowUnusedFunctionParameters" value="true"/>
+		</properties>
+	</rule>
```
Comment (swissspidy) [inline phpcs.xml.dist]:
```suggestion
		</rule>
```

File: phpcs.xml.dist [hunk 2]
```diff
@@ -10,6 +10,11 @@
 	<rule ref="WordPress.WP.I18n"/>
 	<config name="text_domain" value="gutenberg,default"/>
 
+	<rule ref="VariableAnalysis.CodeAnalysis.VariableAnalysis">
+		<properties>
+			<property name="allowUnusedFunctionParameters" value="true"/>
```
Comment (swissspidy) [inline phpcs.xml.dist]: What about using `allowUnusedParametersBeforeUsed` instead?

Comment (jorgefilipecosta) [reply phpcs.xml.dist]: Hi @swissspidy, the rule was causing problems when we need in a class to overload a function defined in the base class without necessarily using all the parameters.
But on second thought these cases are rare and we should not optimize for them, so I followed your suggestion and disabled the rule on the specific cases where it was problematic.

**Review by draganescu — APPROVED** [review]

File: lib/class-wp-rest-block-directory-controller.php [hunk 3]
```diff
@@ -71,10 +71,12 @@ public function register_routes() {
 	 *
 	 * @since 6.5.0
 	 *
-	 * @param WP_REST_Request $request Full details about the request.
 	 * @return WP_Error|bool True if the request has permission, WP_Error object otherwise.
+	 *
+	 * This function is overloading a function defined in WP_REST_Controller so it should have the same parameters.
```
Comment (talldan) [inline lib/class-wp-rest-block-directory-controller.php]: Little confused by the comment as the `$request` param has been removed in the pr. 

I also couldn't see `permissions_check` defined in WP_REST_Controller. Instead there are individual methods (`get_items_permissions_check`, `get_item_permissions_check`, `create_item_permissions_check` ... ).

Comment (jorgefilipecosta) [reply lib/class-wp-rest-block-directory-controller.php]: Nice catch the disable is not needed here I removed it 👍

**Pull request merged**